### PR TITLE
Rewrite memoization logic

### PR DESCRIPTION
**Pull request by disconnect3d**
An attempt to make memoization implementation:
1) with fixed bugs -- current memoization does not handle kwargs 2) maybe more efficient -- current memoization logic will create cache dictionaries per event even if a function is decorated with multiple `reset_on_*` functions; this new impl fixes tha
3) more readable -- hopefully :)

I'll include some simple benchmarking in comments to this PR.

#1453

### CODE REVIEW

File: pwndbg/gdblib/hooks.py [hunk 1]
```diff
@@ -79,7 +79,7 @@ def memoize_on_new_objfile() -> None:
 
 @pwndbg.gdblib.events.start
 def memoize_on_start() -> None:
-    while_running._start_caching()
```
Comment (disconnect3d) [inline pwndbg/gdblib/hooks.py:82]: We never use while_running at this moment, so I commented it for now.

File: pwndbg/lib/memoize.py [hunk 2]
```diff
@@ -12,173 +12,83 @@
...
-
-    _reset = __reset_while_running
+debug = 0
+
+# function -> function wrapper
+wrappers = {}
+functions_to_reset_on = {}
+
+_kwargs_separator = object()
+_sentinel = object()
```
Comment (disconnect3d) [inline pwndbg/lib/memoize.py:22]: Those globals needs better descriptions/comments

**Comment by disconnect3d** [comment]
Some benchmarks below

Without this commit/PR - on current version
```
dc@jhtc:~/pwndbg$ PWNDBG_PROFILE=1 gdb --quiet --nx --ex 'source ./gdbinit.py' /bin/sh --ex 'version' --ex 'entry'
Reading symbols from /bin/sh...
(No debugging symbols found in /bin/sh)
pwndbg: loaded 135 pwndbg commands and 49 shell commands. Type pwndbg [--shell | --all] [filter] for a list.
pwndbg: created $rebase, $ida GDB functions (can be used with print/break)
Time Elapsed: 0.7573239803314209
Gdb:      10.2
Python:   3.6.9 (default, Nov 25 2022, 14:10:45)  [GCC 8.4.0]
Pwndbg:   1.1.1 build: 85e5513
Capstone: 4.0.1024
Unicorn:  2.0.7
Temporary breakpoint 1 at 0x555555558a20

Temporary breakpoint 1, 0x0000555555558a20 in ?? ()
------- tip of the day (disable with set show-tips off) -------
Disable Pwndbg context information display with set context-sections ''
LEGEND: STACK | HEAP | CODE | DATA | RWX | RODATA
─────────────────────────────────────────────────────[ REGISTERS / show-flags off / show-compact-regs off ]─────────────────────────────────────────────────────
*RAX  0x1c
 RBX  0x0
*RCX  0x4
*RDX  0x7ffff7de3b40 (_dl_fini) ◂— push rbp
*RDI  0x7
*RSI  0x7ffff7ffe700 ◂— 0x0
*R8   0x2
 R9   0x0
 R10  0x0
*R11  0x6
*R12  0x555555558a20 ◂— xor ebp, ebp
*R13  0x7fffffffe100 ◂— 0x1
 R14  0x0
 R15  0x0
 RBP  0x0
 RSP  0x7fffffffe100 ◂— 0x1
*RIP  0x555555558a20 ◂— xor ebp, ebp
──────────────────────────────────────────────────────────────[ DISASM / x86-64 / set emulate on ]──────────────────────────────────────────────────────────────
 ► 0x555555558a20    xor    ebp, ebp
   0x555555558a22    mov    r9, rdx
   0x555555558a25    pop    rsi
   0x555555558a26    mov    rdx, rsp
   0x555555558a29    and    rsp, 0xfffffffffffffff0
   0x555555558a2d    push   rax
   0x555555558a2e    push   rsp
   0x555555558a2f    lea    r8, [rip + 0x110ca]
   0x555555558a36    lea    rcx, [rip + 0x11053]
   0x555555558a3d    lea    rdi, [rip - 0x1f4]
   0x555555558a44    call   qword ptr [rip + 0x21858e]    <__libc_start_main>
───────────────────────────────────────────────────────────────────────────[ STACK ]────────────────────────────────────────────────────────────────────────────
00:0000│ r13 rsp 0x7fffffffe100 ◂— 0x1
01:0008│         0x7fffffffe108 —▸ 0x7fffffffe3ae ◂— 0x68732f6e69622f /* '/bin/sh' */
02:0010│         0x7fffffffe110 ◂— 0x0
03:0018│         0x7fffffffe118 —▸ 0x7fffffffe3b6 ◂— 'LC_ALL=en_US.UTF-8'
04:0020│         0x7fffffffe120 —▸ 0x7fffffffe3c9 ◂— 0x524f4c4f435f534c ('LS_COLOR')
05:0028│         0x7fffffffe128 —▸ 0x7fffffffe9b5 ◂— 'SSH_CONNECTION=**redacted**'
06:0030│         0x7fffffffe130 —▸ 0x7fffffffe9ea ◂— 'LESSCLOSE=/usr/bin/lesspipe %s %s'
07:0038│         0x7fffffffe138 —▸ 0x7fffffffea0c ◂— '_=/usr/bin/gdb'
─────────────────────────────────────────────────────────────────────────[ BACKTRACE ]──────────────────────────────────────────────────────────────────────────
 ► f 0   0x555555558a20
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Time Elapsed: 0.16257381439208984
pwndbg> pi from guppy import hpy; h=hpy(); print(h.heap())
Partition of a set of 462751 objects. Total size = 50909192 bytes.
 Index  Count   %     Size   % Cumulative  % Kind (class / dict of class)
     0 111415  24 11121122  22  11121122  22 str
     1 118377  26  8991624  18  20112746  40 tuple
     2  11596   3  4602280   9  24715026  49 dict (no owner)
     3  42237   9  3086506   6  27801532  55 bytes
     4  21002   5  3035576   6  30837108  61 types.CodeType
     5  21032   5  2860352   6  33697460  66 function
     6   2829   1  2745216   5  36442676  72 type
     7   1042   0  2495368   5  38938044  76 dict of module
     8   2829   1  1350424   3  40288468  79 dict of type
     9  10349   2   966752   2  41255220  81 list
<822 more rows. Type e.g. '_.more' to view.>
pwndbg>
```

With this commit/PR:
```
dc@jhtc:~/pwndbg$ PWNDBG_PROFILE=1 gdb --quiet --nx --ex 'source ./gdbinit.py' /bin/sh --ex 'version' --ex 'entry'
Reading symbols from /bin/sh...
(No debugging symbols found in /bin/sh)
pwndbg: loaded 135 pwndbg commands and 49 shell commands. Type pwndbg [--shell | --all] [filter] for a list.
pwndbg: created $rebase, $ida GDB functions (can be used with print/break)
Time Elapsed: 0.7470626831054688
Gdb:      10.2
Python:   3.6.9 (default, Nov 25 2022, 14:10:45)  [GCC 8.4.0]
Pwndbg:   1.1.1 build: 9835737
Capstone: 4.0.1024
Unicorn:  2.0.7
Temporary breakpoint 1 at 0x555555558a20

Temporary breakpoint 1, 0x0000555555558a20 in ?? ()
------- tip of the day (disable with set show-tips off) -------
Use GDB's dprintf command to print all calls to given function. E.g. dprintf malloc, "malloc(%p)\n", (void*)$rdi will print all malloc calls
LEGEND: STACK | HEAP | CODE | DATA | RWX | RODATA
─────────────────────────────────────────────────────[ REGISTERS / show-flags off / show-compact-regs off ]─────────────────────────────────────────────────────
*RAX  0x1c
 RBX  0x0
*RCX  0x4
*RDX  0x7ffff7de3b40 (_dl_fini) ◂— push rbp
*RDI  0x7
*RSI  0x7ffff7ffe700 ◂— 0x0
*R8   0x2
 R9   0x0
 R10  0x0
*R11  0x6
*R12  0x555555558a20 ◂— xor ebp, ebp
*R13  0x7fffffffe100 ◂— 0x1
 R14  0x0
 R15  0x0
 RBP  0x0
 RSP  0x7fffffffe100 ◂— 0x1
*RIP  0x555555558a20 ◂— xor ebp, ebp
──────────────────────────────────────────────────────────────[ DISASM / x86-64 / set emulate on ]──────────────────────────────────────────────────────────────
 ► 0x555555558a20    xor    ebp, ebp
   0x555555558a22    mov    r9, rdx
   0x555555558a25    pop    rsi
   0x555555558a26    mov    rdx, rsp
   0x555555558a29    and    rsp, 0xfffffffffffffff0
   0x555555558a2d    push   rax
   0x555555558a2e    push   rsp
   0x555555558a2f    lea    r8, [rip + 0x110ca]
   0x555555558a36    lea    rcx, [rip + 0x11053]
   0x555555558a3d    lea    rdi, [rip - 0x1f4]
   0x555555558a44    call   qword ptr [rip + 0x21858e]    <__libc_start_main>
───────────────────────────────────────────────────────────────────────────[ STACK ]────────────────────────────────────────────────────────────────────────────
00:0000│ r13 rsp 0x7fffffffe100 ◂— 0x1
01:0008│         0x7fffffffe108 —▸ 0x7fffffffe3ae ◂— 0x68732f6e69622f /* '/bin/sh' */
02:0010│         0x7fffffffe110 ◂— 0x0
03:0018│         0x7fffffffe118 —▸ 0x7fffffffe3b6 ◂— 'LC_ALL=en_US.UTF-8'
04:0020│         0x7fffffffe120 —▸ 0x7fffffffe3c9 ◂— 0x524f4c4f435f534c ('LS_COLOR')
05:0028│         0x7fffffffe128 —▸ 0x7fffffffe9b5 ◂— 'SSH_CONNECTION=**redacted**'
06:0030│         0x7fffffffe130 —▸ 0x7fffffffe9ea ◂— 'LESSCLOSE=/usr/bin/lesspipe %s %s'
07:0038│         0x7fffffffe138 —▸ 0x7fffffffea0c ◂— '_=/usr/bin/gdb'
─────────────────────────────────────────────────────────────────────────[ BACKTRACE ]──────────────────────────────────────────────────────────────────────────
 ► f 0   0x555555558a20
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Time Elapsed: 0.16073989868164062
pwndbg> pi from guppy import hpy; h=hpy(); print(h.heap())
Partition of a set of 463194 objects. Total size = 50945210 bytes.
 Index  Count   %     Size   % Cumulative  % Kind (class / dict of class)
     0 111395  24 11119690  22  11119690  22 str
     1 118514  26  9000048  18  20119738  39 tuple
     2  11601   3  4607672   9  24727410  49 dict (no owner)
     3  42215   9  3085500   6  27812910  55 bytes
     4  20991   5  3034008   6  30846918  61 types.CodeType
     5  21162   5  2878032   6  33724950  66 function
     6   2819   1  2733632   5  36458582  72 type
     7   1042   0  2495368   5  38953950  76 dict of module
     8   2819   1  1346464   3  40300414  79 dict of type
     9  10349   2   966752   2  41267166  81 list
<806 more rows. Type e.g. '_.more' to view.>
pwndbg>
```

File: pwndbg/lib/memoize.py [hunk 3]
```diff
@@ -12,173 +12,83 @@
...
-        while_running.caching = True
-
-    @staticmethod
-    def __reset_while_running() -> None:
-        for obj in while_running.caches:
-            obj.clear()
-        while_running.caching = False
-
-    _reset = __reset_while_running
+debug = 0
```
Comment (gsingh93) [inline pwndbg/lib/memoize.py:15]: Just a thought, this could be an env variable.

File: pwndbg/lib/memoize.py [hunk 4]
```diff
@@ -12,173 +12,83 @@
...
+
+# function -> function wrapper
+wrappers = {}
```
Comment (gsingh93) [inline pwndbg/lib/memoize.py:18]: Would appreciate some type hints for globals like this (doesn't need to be perfect, even `Dict[Callable, Callable]` is fine if it's not easy to be more accurate).

**Review by gsingh93 — COMMENTED** [review]
It'd be nice to have some more comments explaining how this works, as it's pretty complex. Also needs to be rebased.

**Comment by gsingh93** [comment]
What's the plan for this?

**Comment by disconnect3d** [comment]
Update & merge, or, reiterate and rewrite further

**Comment by disconnect3d** [comment]
E.g. we could rewrite the interface to have a single decorator like `@cache_on(start, stop, ...)` instead of multiple.

**Comment by disconnect3d** [comment]
Another attempt in implementing this is below:
```py
"""
Caches return values until some event in the inferior happens,
e.g. execution stops because of a SIGINT or breakpoint, or a
new library/objfile are loaded, etc.
"""

from functools import wraps
from typing import Any
from typing import Callable
from typing import Dict
from typing import Tuple

debug = 0


class _CacheUntilEvent:
    def __init__(self, event):
        self.caches = []
        if event is not None:
            event(self.clear)  # Hook up clear method to the given event

    def clear(self):
        for cache in self.caches:
            cache.clear()
    
    def add_cache(self, cache):
        self.caches.append(cache)

# This is not used anywhere, but it might be useful in the future.
# To make it work, the `IS_CACHING` needs to be per-cache, not global.
# To implement that, we need to:
# 1. Change the `inner` decorator below to pass `func` to the event
# 2. Set `func.is_caching=True` in `decorator` below
# 3. Use `if func.is_caching` instead of `if IS_CACHING` in `decorator`
# All this would allow the `_CacheWhileRunning` to set the 
# `is_caching` status based on start/exit events.
#class _CacheWhileRunning(_CacheUntilEvent):
#    def __init__(self, event_start, event_exit):
#        event_start(self.__start_caching)
#        super().__init__(event_exit)
#        self.func = None
#
#    def add_cache(self, cache, func):
#        super().add_cache(cache)
#        self.func = func
#
#    @pwndbg.gdblib.events.start
#    @staticmethod
#    def __start_caching():
#        self.func.is_caching = True
#
#    def clear(self):
#        super().clear()
#        self.func.is_caching = False


# This needs to be registered using the `register_caches_until_events` function
# which needs to pass an iterable of tuples (event_name, CacheUntilEvent objects)
_ALL_CACHE_UNTIL_EVENTS = {
    "forever": _CacheUntilEvent(None),
}

def register_caches_until_events(tuples):
    for event_name, cache_obj in tuples:
        _ALL_CACHE_UNTIL_EVENTS[event_name] = cache_obj


# A singleton used to mark a cache miss
_NOT_FOUND_IN_CACHE = object()
_KWARGS_SEPARATOR = object()

# Global value whether the results from cache are returned or not
IS_CACHING = True


def cache_until(*event_names) -> Callable:
    def inner(func):
        cache: Dict[Tuple[Any], Any] = {}
        @wraps(func)
        def decorator(*a, **kw):
            if IS_CACHING:
                key: Tuple[Any] = (a, _KWARGS_SEPARATOR, *kw.items())
                value = cache.get(key, _NOT_FOUND_IN_CACHE)
                if value is not _NOT_FOUND_IN_CACHE:
                    return value

                value = func(*a, **kw)

                # Sanity check; should we maybe do it only in debug?
                # TODO/FIXME: Should we do similar sanity check for args/kwargs?
                if instance(value, list):
                    print(f"Should not cache mutable types! {func.__name__}")

                cache[key] = value

                return value

            return func(*a, **kw)

        decorator.cache = cache

        # Register the cache for the given event so it can be cleared
        for event_name in event_names:
            _ALL_CACHE_UNTIL_EVENTS[event_name].add_cache(cache)

        return decorator

    return inner


def clear_caches() -> None:
    for cache in _ALL_CACHE_UNTIL_EVENTS.values():
        cache.clear()
```

Then someone needs to call:
```py
    import pwndbg.lib.cache
    pwndbg.lib.cache.register_caches_until_events((
        ("stop", _CacheUntilEvent(pwndbg.gdblib.events.stop)),
        ("exit", _CacheUntilEvent(pwndbg.gdblib.events.exit)),
        ("objfile", _CacheUntilEvent(pwndbg.gdblib.events.new_objfile)),
        ("start", _CacheUntilEvent(pwndbg.gdblib.events.start)),
        ("cont", _CacheUntilEvent(pwndbg.gdblib.events.cont)),
        ("thread", _CacheUntilEvent(pwndbg.gdblib.events.thread)),
        ("prompt", _CacheUntilEvent(pwndbg.gdblib.events.before_prompt)),
        #"while_running": _CacheWhileRunning(pwndbg.gdblib.events.start, pwndbg.gdbliv.events.exit),
    ))
```
But when I try to put this e.g. in `load_gdblib` function this creates a problem with initialization order:
- gdbinit.py imports color, color imports theme, it imports `from pwndbg.gdblib import config`
- this imports `from pwndbg.gdblib import arch as arch_mod`
- which then goes over some imports up until it ends up in
- gdblib/remote.py which does want to use cache
- but hey, we didn't finish importing stuff in the first place so cache events are not registered yet... :(

This also requires:
```sh
sed -i 's/import pwndbg.lib.memoize/import pwndbg.lib.cache/g' $(find . -type f)
sed -i 's/pwndbg.lib.memoize.reset_on_objfile/pwndbg.lib.cache.cache_until("objfile")/g' $(find . -type f)
sed -i 's/pwndbg.lib.memoize.reset_on_stop/pwndbg.lib.cache.cache_until("stop")/g' $(find . -type f)
sed -i 's/pwndbg.lib.memoize.reset_on_exit/pwndbg.lib.cache.cache_until("exit")/g' $(find . -type f)
sed -i 's/pwndbg.lib.memoize.reset_on_prompt/pwndbg.lib.cache.cache_until("prompt")/g' $(find . -type f)
sed -i 's/pwndbg.lib.memoize.reset_on_thread/pwndbg.lib.cache.cache_until("thread")/g' $(find . -type f)
sed -i 's/pwndbg.lib.memoize.reset_on_start/pwndbg.lib.cache.cache_until("start")/g' $(find . -type f)
sed -i 's/pwndbg.lib.memoize.reset_on_cont/pwndbg.lib.cache.cache_until("cont")/g' $(find . -type f)
sed -i 's/pwndbg.lib.memoize.forever/pwndbg.lib.cache.cache_until("forever")/g' $(find . -type f)
```

**Comment by disconnect3d** [comment]
An alternative implementation was merged in https://github.com/pwndbg/pwndbg/pull/1671

**Pull request closed without merging**